src/OpenTK.Windowing.GraphicsLibraryFramework/JoystickState.cs

```diff
@@ -0,0 +1,235 @@
+
+            return (b0 & pow) == pow && (b1 & pow) != pow;
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void SetButtonDown(int index, bool value)
+        {
+            int byteOffSet = index / 8;
+            int bitOffset = index % 8;
+
+            if (value)
+            {
+                _buttons[_current][byteOffSet] |= (byte)(1 << bitOffset);
+            }
+            else
+            {
+                _buttons[_current][byteOffSet] &= (byte)~(1 << bitOffset);
+            }
+        }
+
+        /// <inheritdoc />
+        public float GetAxis(int index)
+        {
+            return _axes[_current][index];
+        }
+
+        /// <inheritdoc />
+        public float GetAxisPrevious(int index)
+        {
+            return _axes[_previous][index];
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void SetAxis(int index, float value)
+        {
+            _axes[_current][index] = value < -1 ? -1 : (value > 1 ? 1 : value);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void SetAxes(float[] axes)
+        {
+            _axes[_current] = axes;
+        }
+
+        /// <inheritdoc />
+        public override int GetHashCode()
+        {
+            int hashCode = _axes.GetHashCode();
+            hashCode ^= _hats.GetHashCode();
+            hashCode ^= _buttons.GetHashCode();
+            hashCode ^= Id.GetHashCode();
+            hashCode ^= Name.GetHashCode();
+
+            return hashCode;
+        }
+
+        /// <inheritdoc/>
+        public override string ToString()
+        {
+            var builder = new StringBuilder();
+
+            builder.Append("{hats: [");
+            builder.Append(_hats[_current][0]);
+            for (int i = 1; i < _hats.Length; i++)
+            {
+                builder.Append(", ");
+                builder.Append(_hats[_current][i]);
+            }
+
+            builder.Append("], axes: [");
+            builder.Append(_axes[_current][0]);
+            for (int i = 1; i < _axes.Length; i++)
+            {
+                builder.Append(", ");
+                builder.Append(_axes[_current][i]);
+            }
+
+            builder.Append("], buttons: [");
+            builder.Append(IsButtonDown(0) ? "down" : "up");
+            for (int i = 0; i < _buttons.Length * 8; i++)
+            {
+                builder.Append(", ");
+                builder.Append(IsButtonDown(i) ? "down" : "up");
+            }
+
+            builder.Append("], id: ");
+            builder.Append(Id);
+            builder.Append(", name: ");
+            builder.Append(Name);
+            builder.Append("}");
+
+            return builder.ToString();
+        }
+
+        internal unsafe void Update()
+        {
+            _previous = _current;
+            _current++;
+            _current %= MaxHistory;
+
+            var h = GLFW.GetJoystickHatsRaw(Id, out var count);
```

Comment:
Does this allocate?

Reply:
@NogginBops is going to have a play with the GLFW calls separately.